core/src/Cache/Keys/MsalAccessTokenCacheKey.cs

```diff
@@ -0,0 +1,167 @@
+    /// </summary>
+    internal class MsalAccessTokenCacheKey
+    {
+        private readonly string _environment;
+        private readonly string _homeAccountId;
+        private readonly string _clientId;
+        private readonly string _scopes;
+        private readonly string _tenantId;
+
+        internal MsalAccessTokenCacheKey(
+            string environment,
+            string tenantId,
+            string userIdentifier,
+            string clientId,
+            string scopes)
+        {
+            if (string.IsNullOrEmpty(environment))
+            {
+                throw new ArgumentNullException(nameof(environment));
+            }
+
+            if (string.IsNullOrEmpty(clientId))
+            {
+                throw new ArgumentNullException(nameof(clientId));
+            }
+
+           _environment = environment;
+           _homeAccountId = userIdentifier;
+           _clientId = clientId;
+           _scopes = scopes;
+           _tenantId = tenantId;
+        }
+
+
+        public override string ToString()
+        {
+            var stringBuilder = new StringBuilder();
+            stringBuilder.Append((_homeAccountId ?? "") + MsalCacheConstants.CacheKeyDelimiter);
+            stringBuilder.Append(this._environment + MsalCacheConstants.CacheKeyDelimiter);
+            stringBuilder.Append(MsalCacheConstants.AccessToken + MsalCacheConstants.CacheKeyDelimiter);
+            stringBuilder.Append(_clientId + MsalCacheConstants.CacheKeyDelimiter);
+            stringBuilder.Append((_tenantId ?? "") + MsalCacheConstants.CacheKeyDelimiter);
+            stringBuilder.Append(_scopes);
+
+            return stringBuilder.ToString();
+        }
+
+
+        #region UWP
+
+        /// <summary>
+        /// Gets a key that is smaller than 255 characters, which is a limitation for 
+        /// UWP storage. This is done by hashing the scopes and env.
+        /// </summary>
+        /// <remarks>
+        /// accountId - two guids plus separator - 73 chars        
+        /// "accesstoken" string - 11 chars
+        /// clientid - a guid - 36 chars
+        /// tenantid - a guid - 36 chars
+        /// scopes and env - a sha256 string - 44 chars
+        /// delimiters - 4 chars
+        /// total: 204 chars
+        /// </remarks>
+        public string GetFixedSizeKey()
+        {
+            var stringBuilder = new StringBuilder();
+            stringBuilder.Append((_homeAccountId ?? "") + MsalCacheConstants.CacheKeyDelimiter);
+            
+            stringBuilder.Append(MsalCacheConstants.AccessToken + MsalCacheConstants.CacheKeyDelimiter);
+            stringBuilder.Append(_clientId + MsalCacheConstants.CacheKeyDelimiter);
+            stringBuilder.Append((_tenantId ?? "") + MsalCacheConstants.CacheKeyDelimiter);
+
+            // can't use scopes and env because they are of variable length
+            stringBuilder.Append(CoreCryptographyHelpers.CreateSha256Hash(_scopes + _environment));
```

Comment:
Thanks for bringing this up, because the current implementation is not in line with your agreement. The scopes are lowercased and ordered. Assuming the iOS version has shipped, this means we are incompatible :(. We should follow this up separately. 

---
In reply to: [221718990](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1287#discussion_r221718990) [](ancestors = 221718990)

Reply:
We have found that inconsistency during the compatibility testing before and Petro fixed it, if I remember correctly. Let's follow up on that.